besu/src/main/java/org/hyperledger/besu/controller/TargetingGasLimitCalculator.java

```diff
@@ -52,21 +64,32 @@ public long nextGasLimit(final long currentGasLimit) {
 
   @Override
   public void changeTargetGasLimit(final Long targetGasLimit) {
-    checkArgument(targetGasLimit >= 0, "Target gas limit must be non-negative");
+    checkArgument(
+        targetGasLimit >= minGasLimit,
+        "targetGasLimit of " + targetGasLimit + " is below the minGasLimit of " + minGasLimit);
+
+    checkArgument(
+        targetGasLimit <= maxGasLimit,
+        "targetGasLimit of " + targetGasLimit + " is above the maxGasLimit of " + maxGasLimit);
     this.targetGasLimit = targetGasLimit;
   }
 
+  private long adjustAmount(final long currentGasLimit) {
+    final long maxAdjustAmount = Math.max(deltaBound(currentGasLimit) - 1, 0);
+    return Math.min(adjustFactor, maxAdjustAmount);
+  }
+
   private long safeAdd(final long gasLimit) {
     try {
-      return Math.addExact(gasLimit, ADJUSTMENT_FACTOR);
+      return Math.addExact(gasLimit, adjustAmount(gasLimit));
     } catch (final ArithmeticException ex) {
       return Long.MAX_VALUE;
     }
   }
 
   private long safeSub(final long gasLimit) {
     try {
-      return Math.max(Math.subtractExact(gasLimit, ADJUSTMENT_FACTOR), 0);
+      return Math.max(Math.subtractExact(gasLimit, adjustAmount(gasLimit)), 0);
     } catch (final ArithmeticException ex) {
       return 0;
```

Comment:
I think there's a bug here we didn't catch. We don't use the lower and upper bounds in these adds. If that's the case, we should add tests to that effect. One thing which might make the code look closer to what we're going after conceptually is the use of google guava's `Range` class.

Reply:
@RatanRSur  I make the check in `changeTargetGasLimit` method. So it throws exception if people set the targetGasLimit out of the range in constructor (L67 - L73).
Or you prefer another way that allowing to set the `targetGasLimit` below/above the range in constructor while `nextGasLimit` never go out of range.

Reply:
I see, that makes sense!